File: java/revenj-core/src/main/java/org/revenj/serialization/Serialization.java
package org.revenj.serialization;

import org.revenj.Utils;

import java.io.IOException;
import java.lang.reflect.Type;

public interface Serialization<TFormat> {
	TFormat serialize(Type type, Object value) throws IOException;

	default TFormat serialize(Object value) throws IOException {
		return serialize(null, value);
	}

	Object deserialize(Type type, TFormat data) throws IOException;

	@SuppressWarnings("unchecked")
	default <T> T deserialize(TFormat data, Class<T> manifest) throws IOException {
		return (T) deserialize(manifest, data);
	}

	@SuppressWarnings("unchecked")
	default <T> T deserialize(TFormat data, Class<T> container, Type argument, Type... arguments) throws IOException {
		return (T) deserialize(Utils.makeGenericType(container, argument, arguments), data);
	}
}

File: scala/revenj-core/src/main/scala/net/revenj/serialization/Serialization.scala
package net.revenj.serialization

import java.lang.reflect.Type

import net.revenj.Utils

import scala.reflect.runtime.universe.TypeTag
import scala.util.Try

trait Serialization[TFormat] {
  def serialize[T: TypeTag](value: T): Try[TFormat]
  private[revenj] def serializeRuntime(value: Any, manifest: Type): Try[TFormat]
  private[revenj] def serializeRuntime(value: Any): Try[TFormat] = {
    serializeRuntime(value, value.getClass)
  }
  def deserialize[T: TypeTag](input: TFormat): Try[T]
  private[revenj] def deserializeRuntime[T](input: TFormat, manifest: Type): Try[T]
  private[revenj] def deserializeRuntime[T](input: TFormat, container: Class[T], argument: Type, arguments: Type*): Try[T] = {
    deserializeRuntime[T](input, Utils.makeGenericType(container, argument, arguments: _*))
  }
}
